src/etlng/impl/Extraction.cpp

```diff
@@ -0,0 +1,227 @@
+    Copyright (c) 2024, the clio developers.
+
+    Permission to use, copy, modify, and distribute this software for any
+    purpose with or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL,  DIRECT,  INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include "etlng/impl/Extraction.hpp"
+
+#include "data/DBHelpers.hpp"
+#include "data/Types.hpp"
+#include "etl/LedgerFetcherInterface.hpp"
+#include "etl/impl/LedgerFetcher.hpp"
+#include "etlng/Models.hpp"
+#include "util/Assert.hpp"
+#include "util/LedgerUtils.hpp"
+#include "util/Profiler.hpp"
+#include "util/log/Logger.hpp"
+
+#include <google/protobuf/repeated_field.h>
+#include <sys/types.h>
+#include <xrpl/basics/Slice.h>
+#include <xrpl/basics/base_uint.h>
+#include <xrpl/proto/org/xrpl/rpc/v1/get_ledger.pb.h>
+#include <xrpl/proto/org/xrpl/rpc/v1/ledger.pb.h>
+#include <xrpl/protocol/LedgerHeader.h>
+#include <xrpl/protocol/STTx.h>
+#include <xrpl/protocol/Serializer.h>
+#include <xrpl/protocol/TxMeta.h>
+
+#include <algorithm>
+#include <chrono>
+#include <cstddef>
+#include <cstdint>
+#include <iterator>
+#include <memory>
+#include <optional>
+#include <ranges>
+#include <string>
+#include <utility>
+#include <vector>
+
+namespace etlng::impl {
+
+model::Object::ModType
+extractModType(PBModType type)
+{
+    switch (type) {
+        case PBObjType::UNSPECIFIED:
+            return model::Object::ModType::Unspecified;
+        case PBObjType::CREATED:
+            return model::Object::ModType::Created;
+        case PBObjType::MODIFIED:
+            return model::Object::ModType::Modified;
+        case PBObjType::DELETED:
+            return model::Object::ModType::Deleted;
+        default:  // some gRPC system values that we don't care about
+            ASSERT(false, "Tried to extract bogus mod type '{}'", PBObjType::ModificationType_Name(type));
+    }
+
+    std::unreachable();
+}
+
+model::Transaction
+extractTx(PBTxType tx, uint32_t seq)
+{
+    auto raw = std::move(*tx.mutable_transaction_blob());
+    ripple::SerialIter it{raw.data(), raw.size()};
+    ripple::STTx const sttx{it};
+    ripple::TxMeta meta{sttx.getTransactionID(), seq, tx.metadata_blob()};
+
+    static constexpr std::size_t KeySize = 32;
+    std::string keyStr{reinterpret_cast<char const*>(sttx.getTransactionID().data()), KeySize};
```

Comment:
Can ripple::to_string(sttx.getTransactionID()) do the same job?

Reply:
Good question. This code was moved as is from the old version. Let me try